D8AGE-487 Remove the DI container and fix test coverage.

intelektron opened the pull request.

intelektron committed:
D8AGE-487: Remove league container from Composer.

intelektron committed:
D8AGE-487: Introduce ApiFactory.

intelektron committed:
D8AGE-487: Update ApiClient to use factory instead of the container.

intelektron committed:
D8AGE-487: Fix tests to use factory.

intelektron committed:
D8AGE-487: Make the constructor shorter.

intelektron committed:
D8AGE-487: Remove extra endpoint parameter.

intelektron committed:
D8AGE-487: Change the way of supporting token and accessing factory.

intelektron committed:
D8AGE-487: Introduce ApiFactoryInterface.

intelektron committed:
D8AGE-487: Remove endpointUrl from the class configuration.

intelektron committed:
D8AGE-487: Remove logic from DTO.

intelektron committed:
D8AGE-487: Optimize the ClientTestTrait.

intelektron committed:
D8AGE-487: Change links to https to avoid IDE warnings.

intelektron committed:
D8AGE-487: Add simple method tests to ApiClient.

intelektron committed:
D8AGE-487: Small improvements reported by PHPStorm.

intelektron committed:
D8AGE-487: The change in factory tests.

intelektron committed:
D8AGE-487: Update to endpoint tests.

intelektron commented on src/ApiClient.php:
```diff
@@ -33,15 +29,10 @@
  */
 class ApiClient implements ApiClientInterface
 {
-    use ConfigurationAwareTrait;
```
I moved all the configuration/token stuff to the factory so I wouldn't have to pass it back and forth.

intelektron commented on src/ApiClient.php:
```diff
@@ -52,36 +43,33 @@ public function __construct(
+        ), $configuration);
     }
 
     public function requestToken(): Token
```
The interface of this file is not changed.

intelektron commented on src/ApiClient.php:
```diff
@@ -125,86 +113,24 @@ public function getPermanentIdentifier(string $correlationId): string
-    }
-
-    public function setToken(Token $token): self
+    public function setToken(Token $token): ApiClient
```
This may seem a little controversial as it breaks the clean architecture, but on the other hand - it keeps things really simple. The token is passed to the factory class.

intelektron commented on src/Contract/RestInterface.php:
```diff
@@ -23,4 +24,9 @@ public function postJson(string $uri, string $jsonBody, array $headers = []): Re
+    /**
+     * @return array<string, string>
+     */
+    public function getAuthorizationHeadersFromToken(Token $token): array;
```
It was a difficult decision, but I finally moved this code here. Previously it was part of the `TokenAwareTrait`.

intelektron commented on src/Endpoint/AuthorizedEndpointBase.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * Serves as the base for endpoint classes with token support.
+ */
+abstract class AuthorizedEndpointBase extends EndpointBase
```
After removing the container, I noticed that the `TokenAwareTrait` does not make the code any simpler. On the contrary, it requires a lot of extra lines. I changed it to a shorter solution.

intelektron commented on src/Endpoint/EndpointBase.php:
```diff
@@ -30,19 +30,18 @@ abstract class EndpointBase
     /**
      * @param array<string, mixed> $configuration
      */
-    public function __construct(protected RestInterface $rest, string $baseUrl, array $configuration = [])
```
We don't need the separate parameter for base url, it's already in the config.

intelektron commented on src/Endpoint/EndpointBase.php:
```diff
@@ -30,19 +30,18 @@ abstract class EndpointBase
     {
         $resolver = new OptionsResolver();
 
-        $resolver->setRequired('endpointUrl')
```
Validating only `endpointUrl` is a bug, it causes problems if the `baseUrl` is empty.

intelektron commented on src/Endpoint/EndpointBase.php:
```diff
@@ -54,7 +53,7 @@ protected function getConfigurationResolver(): OptionsResolver
      */
     protected function getEndpointUrl(array $replacements = []): string
     {
-        $url = $this->getConfigValue('endpointUrl');
```
Saving this as a separate config value seems overly complicated.

intelektron commented on src/Http/Download.php:
```diff
@@ -5,33 +5,30 @@
     {
         try {
-            $response = $this->rest->get($uri, $this->getAuthorizationHeaders());
+            $response = $this->rest->get($uri, $this->rest->getAuthorizationHeadersFromToken($this->token));
```
So I had to put this method somewhere outside the `AuthorisedEndpointBase`, it is also used by the `Download` class.

drishu replied:
y, I think the trait made sense from this point of view

intelektron replied:
Eventually, I've added a smaller trait that provides this method

intelektron commented on tests/fixtures/json/requests_invalid_request.json:
```diff
@@ -1 +1 @@
-{"departmentCode":"250771","contactUserNames":["DGTRADETUCE"],"deliveryContactUsernames":["DGTRADETUCE"],"phoneNumber":"123456789","title":"Test Title","clientReference":"1","purposeCode":"PC","priorityCode":"SL","deliveryModeCode":"FOOBAR","comments":"Test Comments","referenceSet":{"urls":[{"url":"https:\/\/example.com","shortName":"Example","referenceLanguages":["EN"]}],"files":[{"file":{"fileN ...
\ No newline at end of file
+{"departmentCode":"250771","contactUserNames":["DGTRADETUCE"],"deliveryContactUsernames":["DGTRADETUCE"],"phoneNumber":"123456789","title":"Test Title","clientReference":"1","purposeCode":"PC","priorityCode":"SL","deliveryModeCode":"FOOBAR","comments":"Test Comments","referenceSet":{"urls":[{"url":"https:\/\/example.com","shortName":"Example","referenceLanguages":["EN"]}],"files":[{"file":{"fileN ...
```
These are only changes from `http` to `https`, to make PHPStorm inspections less verbose.

intelektron commented on tests/src/ApiClientTest.php:
```diff
@@ -4,104 +4,157 @@
 /**
  * @coversDefaultClass \OpenEuropa\CdtClient\ApiClient
  */
 class ApiClientTest extends TestCase
```
This was the Achilles heel of the previous version. I've changed it to cover all methods with simple tests. The asserts are not sophisticated, but sufficient to detect any runtime errors.

intelektron commented on tests/src/ApiClientTest.php:
```diff
@@ -4,104 +4,157 @@
-
-    protected function setUp(): void
+    /**
+     * @covers \OpenEuropa\CdtClient\ApiClient
```
This is needed to cover the constructor code too,

intelektron commented on tests/src/ApiClientTest.php:
```diff
@@ -4,104 +4,157 @@
+            new Response(200, [], (string) file_get_contents(__DIR__ . '/../fixtures/json/simple_token_call_response.json')),
+        ];
+        $client = $this->getTestingApiClient([], $responses, false);
+        self::assertInstanceOf(Token::class, $client->requestToken());
```
The assertions do not care whether the expected results are identical, this type of checking is covered by endpoint testing.

intelektron commented on tests/src/Endpoint/IdentifierEndpointTest.php:
```diff
@@ -5,42 +5,35 @@
      */
     public function testIdentifier(string $correlationId, array $clientConfig, array $responses, mixed $expectedResult): void
     {
-        $token = (new Token())->setAccessToken('JWT_TOKEN')
```
I took a microservice-centric approach here. Instead of creating the client and modifying its reflection to access the endpoints illegally, I've just called the constructor of the endpoint. This is shorter and clearer, and makes the test more isolated.

The missing lines are now replaced by `ApiClientTest` and `ApiFactoryTest`.

intelektron commented on tests/src/Traits/ApiTestTrait.php:
```diff
@@ -0,0 +1,96 @@
+ *
+ * Provides helper methods for testing classes that utilize ApiClient and Rest.
+ */
+trait ApiTestTrait
```
Some of this code was part of the `ClientTestTrait` which has been removed. This trait helps us keep the endpoint tests clean.

drishu commented on src/ApiFactory.php:
```diff
@@ -0,0 +1,73 @@
+
+    public function createEndpoint(string $class): EndpointBase
+    {
+        switch ($class) {
```
we could use is_subclass_of and/or implements_interface to target what the class implements/extends so that we don't hardcode these classes

drishu commented on src/Endpoint/AuthorizedEndpointBase.php:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * @inheritDoc
+     */
+    public function __construct(protected RestInterface $rest, array $configuration, protected Token $token)
```
TokenEndpoint is the exception, most if not all other endpoint classes will all be defined the same, so why not add the token parameter to the EndpointBase construct and have the TokenEndpoint class override that construct instead of adding more inheritance.

Of course this means that the comment from the factory class would complicate things but that's where a TokenAwareInterfece would come in handy, lets discuss

intelektron committed:
D8AGE-487: Get rid of AuthorizedEndpointBase.php, move authorization header method.

intelektron committed:
D8AGE-487: Make the factory more elastic.

intelektron committed:
D8AGE-487: Update test coverage.

intelektron commented on src/ApiFactory.php:
```diff
@@ -0,0 +1,74 @@
+
+    public function createEndpoint(string $class): EndpointBase
+    {
+        if ($class === TokenEndpoint::class) {
```
I've added this one to avoid any confusion in the future. Otherwise, PHP will throw an error about invalid configuration, which is misleading.

intelektron commented on src/Traits/AuthorizationHeadersAwareTrait.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public function getAuthorizationHeaders(?Token $token): array
+    {
+        if (null === $token) {
```
This `if` is required by PHPStan. I had a lot of trouble re-implementing this method. In the end, I had to allow nullable tokens in the `Download` class and create additional tests for this exception.

intelektron committed:
D8AGE-487: Remove obsolete use clauses.

intelektron committed:
D8AGE-487: Fix test coverage.

intelektron commented:
I fixed the remaining stuff and added a new test for the `Rest` class. It turned out that 8/12 test paths of the `doRequest' method were not covered.

After my changes there are 4 uncovered paths left. Unfortunately there seems to be a problem with XDebug, explained [here](https://stackoverflow.com/posts/76708811/revisions). We should test each `foreach` with three types of arrays: `iterable`, `empty` and `not iterable` (like null). However, we can only pass `iterable` and `empty` to the `doRequest` method. So the path coverage is incomplete, but there's probably no point in fixing it to skew the test results.

drishu commented on src/ApiClient.php:
```diff
@@ -4,24 +4,20 @@
 use OpenEuropa\CdtClient\Endpoint\ReferenceDataEndpoint;
 use OpenEuropa\CdtClient\Endpoint\RequestsEndpoint;
 use OpenEuropa\CdtClient\Endpoint\StatusEndpoint;
 use OpenEuropa\CdtClient\Endpoint\TokenEndpoint;
```
leftover

intelektron committed:
D8AGE-487: Remove unused `use`.
